src/simple_lottery/algorithm.rs

```diff
@@ -0,0 +1,37 @@
+//! Simple lottery prove and verify functions
+
+use super::proof::Proof;
+use super::setup::Setup;
+use super::types::Hash;
+use crate::utils::sample;
+use crate::utils::types::Element;
+use blake2::{Blake2s256, Digest};
+
+pub(super) fn prove(setup: &Setup, prover_set: &[Element]) -> Option<Proof> {
+    let mut element_sequence = Vec::with_capacity(setup.proof_size as usize);
+    for &element in prover_set {
+        if hash(setup, element) {
+            element_sequence.push(element);
+        }
+        if prover_set.len() as u64 >= setup.proof_size {
+            return Some(Proof { element_sequence });
+        }
+    }
+
+    None
+}
+
+pub(super) fn verify(setup: &Setup, proof: &Proof) -> bool {
+    (proof.element_sequence.len() as u64 == setup.proof_size)
+        && proof
+            .element_sequence
+            .iter()
+            .all(|&element| hash(setup, element))
+}
+
+fn hash(setup: &Setup, element: Element) -> bool {
```

Comment:
Make this function public (or public(super) with an additional function in the wrapper) to check that an individual element can be sent